### PR TITLE
462 improve locale system

### PR DESCRIPTION
**Pull request by AndyMender**
Work in progress for the new `gettext` based translation system. Related to issue #462.

Short interactive test with the `Item` class:
```python
>>># with en_US and en_GB PO files containing translations for the "potion" item "description"
>>> from tuxemon.core.components.locale import T
pygame 1.9.4
Hello from the pygame community. https://www.pygame.org/contribute.html
>>> from tuxemon.core.components.item import Item
>>> import pygame
>>> pygame.init()
(6, 0)
>>> pygame.display.set_mode()
<Surface(1280x1024x32 SW)>
>>> T.locale
'en_US'
>>> potion = Item("potion")
>>># contains the default "en_US" text for item "name" and ":description"
>>> potion.name, potion.description
('Potion', 'Heals a monster by 50 HP.')
>>># load new locale with a different item "description"
>>> T.load_locale("en_GB")
>>> T.locale
'en_GB'
>>># reload item from "slug"
>>> potion.load("potion")   # or potion.load(potion.slug) to make iterating over multiple items easier
>>> potion.name, potion.description
('Potion', 'Restores 50 HP to the monster.')
>>># with the "en_GB" PO file missing the "description" translation
>>> from tuxemon.core.components.locale import T
pygame 1.9.4
Hello from the pygame community. https://www.pygame.org/contribute.html
>>> from tuxemon.core.components.item import Item
>>> import pygame
>>> pygame.init()
(6, 0)
>>> pygame.display.set_mode()
<Surface(1280x1024x32 SW)>
>>> potion = Item("potion")
>>> potion.name, potion.description
('Potion', 'Heals a monster by 50 HP.')
>>># the "en_GB" locale is loaded, but it doesn't have the "description" translation!
>>> T.load_locale("en_GB")
>>> T.locale
'en_GB'
>>> potion.load("potion")
>>># the default text was used, because the matching translated string was not found in the "en_GB" locale PO file
>>> potion.name, potion.description
('Potion', 'Heals a monster by 50 HP.')
```

Main advantages:
* lesser likelihood of game breaking due to missing translations
* simpler automatic or semi-automatic translation thanks to the use POT and PO files and reliance on `gettext`
* lesser reliance on matching JSON fields across locale JSON files

Note that this is a work in progress and many things need to be improved for the system to work fully!

### CODE REVIEW

File: tuxemon/resources/l18n/base.pot [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# SOME DESCRIPTIVE TITLE.
```
Comment (bitcraft) [inline tuxemon/resources/l18n/base.pot]: Is this file still required?

File: tuxemon/resources/l18n/en_US/LC_MESSAGES/base.po [hunk 2]
```diff
@@ -0,0 +1,81 @@
+# SOME DESCRIPTIVE TITLE.
```
Comment (bitcraft) [inline tuxemon/resources/l18n/en_US/LC_MESSAGES/base.po]: This and other PO files can be generated using a script.  I don't think that we would need comments on each `msgid`/`msgstr` pair, just the slug and translated string would be fine.

File: tuxemon/resources/db/item/potion.json [hunk 3]
```diff
@@ -1,5 +1,7 @@
 {
-  "slug": "item_potion",
+  "slug": "potion",
+  "name": "Potion",
```
Comment (bitcraft) [inline tuxemon/resources/db/item/potion.json]: No caps on "Potion"

File: tuxemon/resources/l18n/base.pot (hunk 1 above)
Comment (AndyMender) [reply tuxemon/resources/l18n/base.pot]: Yes and no. This file is the PO template which can be copy-pasted and populated with `slug`s or used directly to generate language-specific PO files.

File: tuxemon/resources/l18n/en_US/LC_MESSAGES/base.po (hunk 2 above)
Comment (AndyMender) [reply tuxemon/resources/l18n/en_US/LC_MESSAGES/base.po]: I thought the comments would be useful, but of course I can remove them :).

File: tuxemon/resources/db/item/potion.json (hunk 3 above)
Comment (AndyMender) [reply tuxemon/resources/db/item/potion.json]: Yes, I still haven't corrected this.

File: tuxemon/resources/l18n/base.pot (hunk 1 above)
Comment (bitcraft) [reply tuxemon/resources/l18n/base.pot]: If its not required by the gettext library, I think any template for translations would either
* Use the English PO
* Use a script to generate template base on slugs in the database and maps

Keeping a template around seems like it would require some upkeep to make sure it is in sync with the database.  I'd like to see it removed if not absolutely needed.

File: tuxemon/resources/l18n/en_US/LC_MESSAGES/base.po (hunk 2 above)
Comment (bitcraft) [reply tuxemon/resources/l18n/en_US/LC_MESSAGES/base.po]: No need to remove them.  Just wanted to make sure you don't feel obligated to keep them, as it could take a lot of time to make comments for each thing.  Slugs should be descriptive enough so that comments are not needed.  Anyway, comments are fine.

**Comment by AndyMender** [comment]
Restoring `WIP` status since most translations are in a broken state and need to be fixed.

**Comment by AndyMender** [comment]
No, I did not. There were way too many changes to the slugs to make it easy enough to write.

File: tuxemon/resources/l18n/es_ES/LC_MESSAGES/base.po [hunk 4]
```diff
@@ -0,0 +1,1209 @@
...
+
+msgid "menu_monster"
+msgstr "Tuxemon"
+
+# TODO: duplicate of menu_monster?
+msgid "menu_monsters"
+msgstr "Monstruos"
+
+msgid "menu_multiplayer"
+msgstr Multijugador"
```
Comment (airon90) [inline tuxemon/resources/l18n/es_ES/LC_MESSAGES/base.po]: Fix the missing `"`

File: tuxemon/resources/l18n/es_ES/LC_MESSAGES/base.po [hunk 5]
```diff
@@ -0,0 +1,1209 @@
...
+## TECHNIQUE TRANSLATIONS ##
+
+msgid "bite"
+msgstr "Mordisco"
+
+msgid "fire_explosion"
+msgstr "Explosión de fuego"
+
+msgid "poison_string"
+msgstr Púa venenosa"
```
Comment (airon90) [inline tuxemon/resources/l18n/es_ES/LC_MESSAGES/base.po]: `"` is missing

File: tuxemon/resources/l18n/es_ES/LC_MESSAGES/base.po [hunk 6]
```diff
@@ -0,0 +1,1209 @@
...
+msgstr "Mono helado"
+
+msgid "taiga"
+msgstr "Taiga"
+
+msgid "horned_raptor"
+msgstr "Rapaz cornuda"
+
+msgid "lions_tooth"
+msgstr Diente de león"
```
Comment (airon90) [inline tuxemon/resources/l18n/es_ES/LC_MESSAGES/base.po]: `"` is missing

**Comment by AndyMender** [comment]
@airon90 Thanks for noticing the bugs. As this is a work-in-progress still, I will probably test-load all translations at the very end to make sure there are no such bugs before lifting the `WIP` flag from the PR.

**Comment by AndyMender** [comment]
I am waiting for some extra stuff from @kerizane (a save game file migration script), but in principle this is finished. 

I updated the en_US locale PO file with the most recent changes and generated one for the it_IT locale as well, since it was completed recently by @airon90. I touched some of the other locales, but I would like to leave this to the individual translators. The locale files are anyhow quite _volatile_ and there is every likelihood that they will keep on changing whilst I commit to my refactoring work.

File: tuxemon/resources/l18n/it_IT/LC_MESSAGES/base.po [hunk 7]
```diff
@@ -0,0 +1,1510 @@
...
+
+
+## DIALOG TRANSLATIONS ##
+
+# general dialogs
+msgid "healmytuxemon"
+msgstr "Vuoi guarire i tuoi Tuxemon?"
+
+msgid "haveagoodday"
+msgstr Ti auguro una buona giornata!"
```
Comment (airon90) [inline tuxemon/resources/l18n/it_IT/LC_MESSAGES/base.po]: Check what you produce! `"` is missing here

**Comment by kerizane** [comment]
AFAIK we're still supporting Python2. Can you check there are no other issues?

  File "/home/k/Tuxemon/tuxemon/core/prepare.py", line 49, in <module>
    os.makedirs(paths.USER_GAME_DIR, exist_ok=True)
TypeError: makedirs() got an unexpected keyword argument 'exist_ok'

**Comment by kerizane** [comment]
Typo here stops the map loading:
https://github.com/Tuxemon/Tuxemon/blob/23632b03168c17a6c49cc046e84f6dc4972a7298/tuxemon/resources/maps/cotton_town.tmx#L404

File: tuxemon/core/prepare.py [hunk 8]
```diff
@@ -34,52 +34,48 @@
...
-BASEDIR = os.path.abspath(os.path.join(os.path.dirname(os.path.realpath(__file__)), "..")) + os.sep
-if "library.zip" in BASEDIR:
-    BASEDIR = os.path.abspath(os.path.join(BASEDIR, "..")) + os.sep
 
-# Set up our config directory
-CONFIG_PATH = get_config_path() + "/.tuxemon/"
+# Create game dir if missing
 try:
-    os.makedirs(CONFIG_PATH)
+    os.makedirs(paths.USER_GAME_DIR, exist_ok=True)
```
Comment (kerizane) [inline tuxemon/core/prepare.py]: We don't need "exist_ok", all it does it break python2. 
The `except OSError` is enough.

Comment (kerizane) [reply tuxemon/core/prepare.py]: Please check that this works for PR works for py2 as well.

**Comment by AndyMender** [comment]
Topic branch for this PR was rebased from `Tuxemon/development` and extra fixes were added. Also, I removed the `WIP` flag.

**Comment by bitcraft** [comment]
So, I see there are also a lot of changes that are out of scope of the ticket.  Unfortunately I don't have time to go through it all.  In the future, please try to stick to the scope of the ticket when making changes.  

Could you justify these changes briefly?
* Changing the translator from `trans` to convention to `T`
* The "constants" module
* Rename "Npc" to "NPC"
* all the changes to paths (using os.path.join)
* changes in platform/\_\_init\_\_.py

These are welcome changes. but not within the scope of the ticket.  I would have much preferred you not change anything except for just the what you said in the ticket.

Just a FYI, in medium sized projects like this, its really best to not touch old code unless you break up your changes and test.

Like I said, I can't really go through this line-by-line, but in the future, I expect much smaller changes.

**Comment by kerizane** [comment]
@bitcraft I'm happy to take a look at it. Wait for a nod from me if you can.

**Comment by kerizane** [comment]
Can you squash your commits?
Each commit should leave the system in a working state. 
I'd say we only need a couple: Path-related refactoring and the new translation system.

**Comment by kerizane** [comment]
You're also bringing in incomplete and deprecated monsters.

I've updated the sync-wiki-monsters.py to only bring in completed monsters, which I think makes more sense.

Update the script to be consistent with your translation system, remove your changes to the monster.json's and rerun the script.

**Comment by AndyMender** [comment]
> Ouch! Major clobbering of the monster files.
> 
> The entries below aren't needed any more:
> 
> {
>   "attack_base":            60,
>   "defense_base":           7,
>   "hp_base":                50,
>   "special_attack_base":    6,
>   "special_defense_base":   10,
>   "speed_base":             7,
>   "attack_mod":             1,
>   "defense_mod":            1,
>   "hp_mod":                 1,
>   "special_attack_mod":     1,
>   "special_defense_mod":    1,
>   "speed_mod":              1,
>   "exp_give_mod":           3,
>   "exp_req_mod":            27,
>   "sprites": {
>     "battle1": "gfx/sprites/battle/aardart-front.png",
>     "battle2": "gfx/sprites/battle/aardart-back.png",
>     "menu1": "gfx/sprites/battle/aardart-menu01.png",
>     "menu2": "gfx/sprites/battle/aardart-menu02.png"
>   },
> }
> 
> You overwrote the movelists and are bringing in incomplete and deprecated monsters.
> 
> I think it's best to leave any whitespace changes of monster.json for another ticket. The change in this ticket should be minimal, ~4 lines.

I don't think I follow. I just did a regular rebase from the `development` branch.

> I've updated the sync-wiki-monsters.py to only bring in completed monsters, which I think makes more sense. If you were using it, please update the script to be consistent with your translation system, remove your changes to the monster.jsons and rerun the script.

Does the updated version of the script still incorporate the `TuxepediaStore` class which I wrote to more gracefully handle sprite scraping and fix some minor bugs?

**Comment by kerizane** [comment]
A few issues:
- eaglace.json was corrupted by the rebase and stops the game launching
- Line 20 in tuxemon/core/states/combat/combat_menus.py needs to be removed.
- Random encounters don't work. You need to update the slugs in /Tuxemon/tuxemon/resources/db/encounter/*.json
- Trainer battles don't start; when you talk to liela in cotton_town she should fight you.
- route1 crashes as  tuxemon/resources/db/npc/xerogrunt1.json is missing.

Dialogue isn't working in some places:
- In cotton_town aeble and allie say the msgid, not the message string. 
- The professor seems to be dropping some dialogue.

**Comment by kerizane** [comment]
> Does the updated version of the script still incorporate the TuxepediaStore class which I wrote to more gracefully handle sprite scraping and fix some minor bugs?

It does

**Comment by bitcraft** [comment]
This has been open for almost 2 moths and is a blocker.  Do you need help merging it?

**Comment by AndyMender** [comment]
> This has been open for almost 2 moths and is a blocker. Do you need help merging it?
I had some personal matters to attend to and could do only rudimentary fixes to the map issues noticed by @kerizane 

I will have more time the following week to do proper game testing and fix the remaining issues. I'm super sorry it's taking so long.

**Comment by bitcraft** [comment]
Cool, thanks for checking in again.

File: tuxemon/core/components/event/actions/random_encounter.py [hunk 9]
```diff
@@ -103,7 +103,7 @@ def start(self):
             current_monster.current_hp = current_monster.hp
 
             # Create an NPC object which will be this monster's "trainer"
-            npc = Npc("npc_maple")
+            npc = NPC("npc_maple")
```
[the remark this answers is not in the document]
Comment (kerizane) [reply tuxemon/core/components/event/actions/random_encounter.py]: Do random encounters still work? The prefix might break it...

**Comment by kerizane** [comment]
I'll try to take a look over the next few days.

File: tuxemon/core/components/monster.py [hunk 10]
```diff
@@ -274,26 +273,67 @@ def load_from_db(self, slug):
...
-        types = results.get("types")
-        if types:
-            self.type1 = results["types"][0].lower()
-            if len(types) > 1:
-                self.type2 = results["types"][1].lower()
+        self.name = T.translate(results["slug"])                # translated name
+        self.description = T.translate("{}_description".format(results["slug"]))  # translated description
+        self.category = T.translate(results["category"])        # translated category
+
+        self.hp = results["hp_base"]
```
[the remark this answers is not in the document]
Comment (AndyMender) [reply tuxemon/core/components/monster.py]: I'll try to compare my branch with the `development` branch by hand then, because I just performed a standard rebase. I had no intention of overwriting your work, apologies.

**Comment by bitcraft** [comment]
@AndyMender Merging is sometimes better than rebasing when changes are really big or your branch is part behind, *especially* when you are making changes to the same code another person is working on.  In this case, merging development into your feature branch would have shown you all the conflicts, while a rebase will silently overwrite them (sometimes useful).

Please don't forget to squash these commits.

**Comment by kerizane** [comment]
> Merging is sometimes better than rebasing ...
My fault for requesting the rebase.

File: tuxemon/core/components/monster.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kerizane) [reply tuxemon/core/components/monster.py]: Sorry for all the pain caused by that rebase I asked for...

Getting there, we don't need these:
  "sprites": {
    "battle1": "gfx/sprites/battle/aardart-front.png",
    "battle2": "gfx/sprites/battle/aardart-back.png",
    "menu1": "gfx/sprites/battle/aardart-menu01.png",
    "menu2": "gfx/sprites/battle/aardart-menu02.png"
  },

You're also changing
"types": ["earth"],
to 
"types": ["Earth"],

It shouldn't break anything though.

**Comment by kerizane** [comment]
Looks like you indent with spaces, whilst we indent with tabs: this breaks python3.
You didn't properly refactor monster.load -> monster.load_from_db: this breaks the prof giving you a starter.

A feature/convention we had before was that putting a series of strings in an array would open a new text box for each one. Your system combines them into one box. You can see this when you talk to the professor.

**Comment by AndyMender** [comment]
> Looks like you indented with tabs in a few places: this breaks python3.

I used a diff program called Meld, instead of regular Python tools and that's probably why. Fixed.

> You didn't properly refactor monster.load -> monster.load_from_db: this breaks the prof giving you a starter.

Not sure why was this renamed, because other classes still use `load` methods. Hopefully it's fixed now as well.

> A feature/convention we had before was that putting a series of strings in an array would open a new text box for each one. Your system combines them into one box. You can see this when you talk to the professor.

Are you talking about the translation system? Yes, as far as I know, now all multi-line translations are converted into a single string with end-line symbols. If this is an issue, I'll try to look for a solution.

**Comment by kerizane** [comment]
I'd say we should keep the output the same. If it's too much trouble, I'm happy for you to split the screens into separate variables and add the "multi-screen" support at a later date.

The feature was added by #423, and I don't think that much has been added since then, so you could do a partial revert of the commit to fix the map scripts.

**Comment by kerizane** [comment]
> **EDIT:** It was possible to reinstate the previous scrolling/paging mechanism for multi-line dialogues, however the new translation system sadly does not support end-line symbols within single translation lines.

When you talk to the professor an additional blank screen is added at the end. I think it's best to strip the final \n by default.

I'm happy with this solution for now.

**Comment by kerizane** [comment]
Nearly there, I can't see anything obviously wrong with it.

Please test the scenarios that bitcraft requested.

**Comment by AndyMender** [comment]
> When you talk to the professor an additional blank screen is added at the end. I think it's best to strip the final \n by default.

I addressed this by stripping dangling end-line characters via `str.rstrip()`.

> Nearly there, I can't see anything obviously wrong with it.
> Please test the scenarios that bitcraft requested.

Super big thanks for your support so far! :) I am currently game-testing and fixing the leftover bugs.

**Comment by kerizane** [comment]
I notice that some of the msgids ("spillit\n") have "\n" after them, does this cause any issues? 

Otherwise, I'm OK with this. 

I initially wasn't too excited about this change, but a quick google gave me https://poedit.net/, which seems useful. Any other cool things @AndyMender?

**Comment by AndyMender** [comment]
> I notice that some of the msgids ("spillit\n") have "\n" after them, does this cause any issues?

Yes and no. This is needed for the `gettext` utility to recognise multi-line translation keys. If editing .tmx Tiled map files by hand, it's enough to place the end-line symbol `\n` at the end of dialog events with such keys. If editing map files directly in Tiled, it may try to escape the end-line symbol like so `\\n`, but that's more of a display gimmick. The map file will contain the correct, non-escaped end-line symbol afterwards.

> I initially wasn't too excited about this change, but a quick google gave me https://poedit.net/, which could be a great improvement for the translators/content creators.

Yes, poedit is a great utility for building translations. I actually used it to verify some of the generated MO files, because apparently Python built-in tools are broken and use ASCII encoding instead of UTF-8 for converting bytes to strings. Obviously, ASCII is not going to work for many languages. `polib` turned out to be broken as well, hence I settled with `babel` for converting between PO and MO files. It's a mature project, used very often in translations.

> Any other cool things we can do with this @AndyMender?

There is an online auto-translation project which generates PO files, called weblate. It could be used for initial machine-based translations to new languages. Also, using POT template files gives a consistent base for all translation files in case new keys are added or old ones removed.

**Comment by AndyMender** [comment]
> New bug:
> 
>     * Talking to the 2 guards and selecting "Yes" results in a crash.  Looks like a map is corrupt.

There was indeed a map corruption issue. PyCharm tried to outsmart me with automatic XML tag-closing and I didn't notice it.

I also fixed a bug with the new `TranslatorPo.format` method and the `replace_text` function in "tuxemon.core.components.event.actions.__init__.py". Apparently, depending on the translation context, the `${{name}}` tag can mean different things (name of a tuxemon, item or player name), which is quite inconsistent.

**Comment by bitcraft** [comment]
@AndyMender Merge conflicts.  Please resolve and I will test.  Lets get this wrapped up please.

**Comment by AndyMender** [comment]
@bitcraft merge conflicts resolved. There were still some typos in the map files, but I fixed them.

**Comment by bitcraft** [comment]
@AndyMender Got this error. 
image

**Comment by bitcraft** [comment]
I've tested this and it seems to be working.  Thanks for all your hard work and keeping up with the request of me and the other team members!  I hope you don't mind if I defer locale issues to you in the future.  Thanks again!

**Comment by bitcraft** [comment]
...reopen and merge because of mis-click...
